src/llms/openai.ts

```diff
@@ -0,0 +1,242 @@
+  /**
+   * A unique identifier representing your end-user, which can help OpenAI to monitor and detect abuse.
+   */
+  user?: string;
+}
+
+interface Choice {
+  message: Message;
+}
+
+interface ChatCompletionsResponse<T = Choice> {
+  choices: T[];
+}
+
+interface ContentMessage {
+  content: string;
+}
+
+interface RoleMessage {
+  role: string;
+}
+
+interface DoneMessage {
+  done: boolean;
+}
+
+type ChatCompletionsDelta = ContentMessage | RoleMessage | DoneMessage;
+
+interface ChatCompletionsChunk {
+  delta: ChatCompletionsDelta;
+}
+
+const isContentMessage = (message: any): message is ContentMessage => {
+  return message.content !== undefined;
+}
+
+const isDoneMessage = (message: any): message is DoneMessage => {
+  return message.done !== undefined;
+}
+
+/**
+ * Make a request to OpenAI's chat-completions API via the Grafana LLM plugin proxy.
+ */
+export async function chatCompletions(request: ChatCompletionsRequest): Promise<string> {
+  const response = await getBackendSrv().post<ChatCompletionsResponse>('/api/plugins/grafana-llm-app/resources/openai/v1/chat/completions', request, {
+    headers: { 'Content-Type': 'application/json' }
+  });
+  return response.choices[0].message.content;
+}
+
+/**
+ * Make a streaming request to OpenAI's chat-completions API via the Grafana LLM plugin proxy.
+ *
+ * A stream of tokens will be returned as an `Observable<string>`. Use rxjs' `scan` if you want
+ * to produce a new stream containing the concatenated tokens so far.
+ *
+ * @example <caption>Example of accumulating tokens in a stream.</caption>
+ * const stream = streamChatCompletions({ model: 'gpt-3.5-turbo', messages: [
+ *   { role: 'system', content: 'You are a great bot.' },
+ *   { role: 'user', content: 'Hello, bot.' },
+ * ]}).pipe(scan((acc, delta) => acc + delta, ''));
+ */
+export function streamChatCompletions(request: ChatCompletionsRequest): Observable<string> {
+  const channel: LiveChannelAddress = {
+    scope: LiveChannelScope.Plugin,
+    namespace: LLM_PLUGIN_ID,
+    path: `/openai/v1/chat/completions`,
+    data: request,
+  };
+  const messages = getGrafanaLiveSrv()
+    .getStream(channel)
+    .pipe(filter((event) => isLiveChannelMessageEvent(event))) as Observable<LiveChannelMessageEvent<ChatCompletionsResponse<ChatCompletionsChunk>>>
+  return messages.pipe(
+    takeWhile((event) => !isDoneMessage(event.message.choices[0].delta)),
+    map((event) => event.message.choices[0].delta),
+    filter((delta) => isContentMessage(delta)),
+    map((delta) => (delta as ContentMessage).content),
+  );
+}
+
+/**
+ * Check if the OpenAI API is enabled via the LLM plugin.
+ */
+export const enabled = async () => {
+  try {
+    const settings = await getBackendSrv().get(`${LLM_PLUGIN_ROUTE}/settings`, undefined, undefined, {
+      showSuccessAlert: false, showErrorAlert: false,
+    });
+    return settings.enabled && (settings?.secureJsonFields?.openAIKey ?? false);
+  } catch (_e) {
```

Comment:
Would handling the error here be helpful at all for any users to trouble shoot possible issues with connecting?

Reply:
Hmm quite possibly yes, perhaps we should just log a warning?

Reply:
I've added some logging in the first time this happens, let me know what you think (especially about stringifying and logging the error 🤔)